File: src/lib/sql/sourceReference.ts
import { qualifiedTable } from '@/lib/sql/identifierQuoting';
import { inferFileFormat, buildFilePath, type FileFormat } from '@/lib/fileFormat';
import type { TableTypeConfig } from '@/types/tableTypeConfig';
import type { MappingRow } from '@/types/mapping';

/** Low-level: a file.`path` reference when file-backed, otherwise the normal quoted table reference. */
export function resolveReference(
  kind: 'table' | 'file',
  schema: string | undefined,
  table: string,
  fileFormat: FileFormat | undefined,
  filePath: string | undefined
): string {
  if (kind === 'file' && filePath) {
    return `${fileFormat ?? 'csv'}.\`${filePath}\``;
  }
  return qualifiedTable(schema, table);
}

/**
 * Resolves the SOURCE-side FROM reference for a table group: auto-detects the file path/format
 * from the group's own rows' `sourceFileLocation`/`sourceFileName` columns when the mapping doc
 * has them, falling back to the per-table manual override collected in the Preview step.
 */
export function resolveSourceReference(
  config: TableTypeConfig,
  rows: MappingRow[],
  sourceSchema: string | undefined,
  sourceTable: string
): string {
  if (config.sourceKind !== 'file') {
    return qualifiedTable(sourceSchema, sourceTable);
  }

  const rowWithFile = rows.find((r) => r.sourceFileLocation || r.sourceFileName);
  const detectedPath = rowWithFile
    ? buildFilePath(rowWithFile.sourceFileLocation, rowWithFile.sourceFileName)
    : '';
  const filePath = detectedPath || config.sourceFilePathOverride || '';

  const detectedFormat = rowWithFile?.sourceFileName ? inferFileFormat(rowWithFile.sourceFileName) : null;
  const fileFormat = detectedFormat ?? config.sourceFileFormatOverride ?? 'csv';

  return resolveReference('file', sourceSchema, sourceTable, fileFormat, filePath);
}

/** Resolves the TARGET-side FROM reference. Only called for targetKind 'table'/'file' -- callers skip 'dashboard' targets before reaching this. */
export function resolveTargetReference(
  config: TableTypeConfig,
  targetSchema: string | undefined,
  targetTable: string
): string {
  if (config.targetKind !== 'file') {
    return qualifiedTable(targetSchema, targetTable);
  }
  const fileFormat = config.targetFileFormatOverride ?? 'csv';
  return resolveReference('file', targetSchema, targetTable, fileFormat, config.targetFilePathOverride || '');
}

File: server/engine/source_reference.py
"""Direct port of src/lib/sql/sourceReference.ts."""

from __future__ import annotations

from engine.file_format import build_file_path, infer_file_format
from engine.identifier_quoting import qualified_table
from engine.models import MappingRow, TableTypeConfig


def resolve_reference(kind: str, schema: str | None, table: str, file_format: str | None, file_path: str | None) -> str:
    if kind == "file" and file_path:
        return f"{file_format or 'csv'}.`{file_path}`"
    return qualified_table(schema, table)


def resolve_source_reference(
    config: TableTypeConfig, rows: list[MappingRow], source_schema: str | None, source_table: str
) -> str:
    if config.source_kind != "file":
        return qualified_table(source_schema, source_table)

    row_with_file = next((r for r in rows if r.source_file_location or r.source_file_name), None)
    detected_path = (
        build_file_path(row_with_file.source_file_location, row_with_file.source_file_name) if row_with_file else ""
    )
    file_path = detected_path or config.source_file_path_override or ""

    detected_format = infer_file_format(row_with_file.source_file_name) if row_with_file and row_with_file.source_file_name else None
    file_format = detected_format or config.source_file_format_override or "csv"

    return resolve_reference("file", source_schema, source_table, file_format, file_path)


def resolve_target_reference(config: TableTypeConfig, target_schema: str | None, target_table: str) -> str:
    if config.target_kind != "file":
        return qualified_table(target_schema, target_table)
    file_format = config.target_file_format_override or "csv"
    return resolve_reference("file", target_schema, target_table, file_format, config.target_file_path_override or "")
